src/snackbar.tsx

```diff
@@ -32,189 +35,291 @@ type Props = {
+
+        return (
+            <SnackbarComponent
+                ref={ref}
+                message={message}
+                duration={duration}
+                buttonText={buttonText}
+                type={type}
+                onClose={onCloseRef.current}
+                withDismiss={withDismiss}
+            />
+        );
+    }
+);
+
+export default Snackbar;
+
+type SnackbarEntry = Props & {
+    id: string;
 };
 
-const Snackbar: React.FC<Props> = ({
-    message,
-    buttonText,
-    duration,
-    onClose: onCloseProp = () => {},
-    type = 'INFORMATIVE',
-    withDismiss,
-}) => {
-    const defaultDuration = buttonText ? DEFAULT_DURATION_WITH_BUTTON : DEFAULT_DURATION_WITHOUT_BUTTON;
-    duration = Math.max(duration ?? defaultDuration, defaultDuration);
+const SnackbarContext = React.createContext<{
+    snackbars: Array<SnackbarEntry>;
+    setSnackbars: React.Dispatch<React.SetStateAction<Array<SnackbarEntry>>>;
+}>({
+    snackbars: [],
+    setSnackbars: () => {},
+});
+
+export const SnackbarRoot = ({children}: {children: React.ReactNode}): JSX.Element => {
+    const [snackbars, setSnackbars] = React.useState<Array<SnackbarEntry>>([]);
+    const snackbarRef = React.useRef<ImperativeHandle & HTMLDivElement>(null);
+    const isClosingRef = React.useRef(false);
     const renderNative = isWebViewBridgeAvailable();
-    const onCloseRef = React.useRef(onCloseProp);
 
     React.useEffect(() => {
-        onCloseRef.current = onCloseProp;
-    }, [onCloseProp]);
+        // multiple snackbars, close the current one
+        if (snackbars.length > 1 && !isClosingRef.current) {
+            isClosingRef.current = true;
+            if (renderNative) {
+                // the native side will automatically close the current snackbar when opening a new one
+                setSnackbars((snackbars) => snackbars.slice(1));
+            } else {
+                snackbarRef.current?.close({action: 'CONSECUTIVE'});
+            }
+        }
+    }, [snackbars, renderNative]);
 
-    React.useEffect(() => {
-        if (renderNative) {
-            // these are the duration values understood by native app, other values will be ignored
-
-            nativeMessage({
-                message,
-                // @ts-expect-error duration can be 'PERSISTENT' in new webview-bridge lib versions, and old apps will ignore it
-                duration: duration === Infinity ? 'PERSISTENT' : undefined,
-                buttonText,
-                type,
-                ...{withDismiss},
-            }).then((result: unknown) => {
-                // there are terser ways to do this checks, but this one satisfies TS
-                if (
-                    !result ||
-                    typeof result !== 'object' ||
-                    !('action' in result) ||
-                    typeof result.action !== 'string' ||
-                    (result.action !== 'DISMISS' && result.action !== 'TIMEOUT' && result.action !== 'BUTTON')
```

Comment:
yes, it was intentional. We were hiding this native case because we didn't have the case in web implementation. But it makes sense to propagate it now that we support it